application/models/Permission.php

```diff
@@ -703,11 +706,20 @@ public function getButtons(): string
      */
     public function hasSurveyPermission($iSurveyID, $sPermission, $sCRUD = 'read', $iUserID = null)
     {
-        $oSurvey = Survey::Model()->findByPk($iSurveyID);
```

Comment:
It's already a static variables

Reply:
> we get:
> 
> ![image](https://private-user-images.githubusercontent.com/2805488/307921723-fda229c2-1406-4169-8221-bc710e89cb63.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMjM3MDgsIm5iZiI6MTcwOTAyMzQwOCwicGF0aCI6Ii8yODA1NDg4LzMwNzkyMTcyMy1mZGEyMjljMi0xNDA2LTQxNjktODIyMS1iYzcxMGU4OWNiNjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMDg0MzI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2YzYjdmMDdlNzkwZTA5ODhkYjk1NGNkZDkwYTI5NDM4ZmNjM2E1NDE0NmExZjIyZDE0M2U2NmUzNzI1YmI3ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tCtxAG0sBlO9wqta2Hh0dGoIyWT2atbZIBdC0R3C9YQ)

Yes, but you pit the echo at the place you want …

> This was the reason behind the change

Yes : i understand this, but since both call use static var, i think there are maybe a better solution to use the static var at a better place in call.

Not in oSurvey, but maybe (surely) in hasPermission directly !

Reply:
@Shnoulle We may need both. I have measured the total time spent in

[1] `Permission::hasPermission` when being triggered from `Permission::hasSurveyPermission` either directly or indirectly.
[2] `Permission::hasSurveyPermission`

I did it cumulatively, so I added each measurement of [1] into one value and [2] into another one. If my caching mechanism was not in place, then the result was

![image](https://github.com/LimeSurvey/LimeSurvey/assets/2805488/e047f193-a493-48ac-a914-e2204b237162)

so 39% of the time spent for survey performance detection happened outside calls for `hasPermission`, ~ 38ms.

With the caching I implemented the former takes next  to nothing and the latter takes ~25ms.

![image](https://github.com/LimeSurvey/LimeSurvey/assets/2805488/13b65789-626a-48fa-b62e-d52bd88183da)

So, the question is whether the problem you mentioned exists. If so, then we will need to remove this improvement. If not, then this improvement is beneficial.

However, you are right about fixing the cache inside `hasPermission`, which is a good idea. Yet, besides of that we win further time by having a caching mechanism in oSurvey as well, as long as we do not cause harm by it. I will ask around about this.